zjit/src/profile.rs

```diff
@@ -79,25 +80,100 @@ fn profile_insn(profiler: &mut Profiler, bare_opcode: ruby_vminsn_type) {
     }
 }
 
+const DISTRIBUTION_SIZE: usize = 4;
+
+pub type TypeDistribution = Distribution<ProfiledType, DISTRIBUTION_SIZE>;
+
+pub type TypeDistributionSummary = DistributionSummary<ProfiledType, DISTRIBUTION_SIZE>;
+
 /// Profile the Type of top-`n` stack operands
 fn profile_operands(profiler: &mut Profiler, profile: &mut IseqProfile, n: usize) {
     let types = &mut profile.opnd_types[profiler.insn_idx];
-    if types.len() <= n {
-        types.resize(n, Empty);
+    if types.is_empty() {
+        types.resize(n, TypeDistribution::new());
     }
     for i in 0..n {
-        let opnd_type = Type::from_value(profiler.peek_at_stack((n - i - 1) as isize));
-        types[i] = types[i].union(opnd_type);
-        if let Some(object) = types[i].gc_object() {
-            unsafe { rb_gc_writebarrier(profiler.iseq.into(), object) };
-        }
+        let obj = profiler.peek_at_stack((n - i - 1) as isize);
+        // TODO(max): Handle GC-hidden classes like Array, Hash, etc and make them look normal or
+        // drop them or something
+        let ty = ProfiledType::new(obj.class_of(), obj.shape_id_of());
+        unsafe { rb_gc_writebarrier(profiler.iseq.into(), ty.class()) };
+        types[i].observe(ty);
+    }
+}
+
+/// opt_send_without_block/opt_plus/... should store:
+/// * the class of the receiver, so we can do method lookup
+/// * the shape of the receiver, so we can optimize ivar lookup
+/// with those two, pieces of information, we can also determine when an object is an immediate:
+/// * Integer + SPECIAL_CONST_SHAPE_ID == Fixnum
+/// * Float + SPECIAL_CONST_SHAPE_ID == Flonum
+/// * Symbol + SPECIAL_CONST_SHAPE_ID == StaticSymbol
+/// * NilClass == Nil
+/// * TrueClass == True
+/// * FalseClass == False
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct ProfiledType {
+    class: VALUE,
+    shape: ShapeId,
```

Comment:
For non-ivar instructions, is there any plan to use any information other than `SPECIAL_CONST_SHAPE_ID`?

I feel like `immediate: bool` is what you actually wanted to use. Object shapes in Ruby are basically a cache key to access ivars today, so using all those 32 bits to store the single flag `SPECIAL_CONST_SHAPE_ID` seems space-inefficient.

For shapes, some instructions would need to query non-`class`/`immediate` information, so it might make sense to prepare a storage for instruction-specific profiling data, and store shapes only for ivar instructions.

Reply:
The tuple of (class, shape) is useful for ivar instructions and also sends --- we may see cases where a send `a.b` is actually a send to an `attr_accessor`, inline that, and then can use the shape information to optimize that into a hypothetical `LoadField`.

Re: other methods: we could absolutely do a discriminated union of shape|bitfield. If we come up with more bits, I would be delighted to do that!

Reply:
Sure, attr accessors are a good point, but the point is that it's still applicable to only a limited number of instructions. Whether you come up with other bits or not, it'd be more space-efficient to make the storage size for profiling data instruction-specific, instead of using the largest number of bits for every instruction.

It's probably too early to optimize for space efficiency now, but from the maintenance point of view, it seems misleading to store shapes (instead of `immediate: bool`) for every instruction when (non-`SPECIAL_CONST_SHAPE_ID`) shapes are used only for 3 out of 110 instructions (getivar, setivar, opt_send_without_block).

Reply:
Yes, though if we do that we either need to keep them aligned (in which case they are all the same size anyway) or do variable-width profiles (which imo requires additional design).

Reply:
I think starting a doc of what profile information we might want per instruction is a great idea. So far I have:

* ivar stuff: (class, shape)
* sends: (class, shape)
* math ops: (class, immediate bit (which we could even fold into the class pointer?))
* opt_nil_p: (bool nilness)
* conditional branch instructions: class